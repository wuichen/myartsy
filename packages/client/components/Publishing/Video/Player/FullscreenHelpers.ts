export const requestFullscreen = e => {
  if (e.requestFullscreen) {
    e.requestFullscreen()
  } else if (e.webkitRequestFullscreen) {
    e.webkitRequestFullscreen()
  } else if (e.mozRequestFullScreen) {
    e.mozRequestFullScreen()
  } else if (e.msRequestFullscreen) {
    e.msRequestFullscreen()
  }
}

export const exitFullscreen = () => {
  if (document.exitFullscreen) {
    document.exitFullscreen()
  } else if (document.webkitExitFullscreen) {
    document.webkitExitFullscreen()
  } else if (document.mozCancelFullScreen) {
    document.mozCancelFullScreen()
  } else if (document.msExitFullscreen) {
    document.msExitFullscreen()
  }
}

export const fullscreenEnabled = () => {
  return (
    document.fullscreenEnabled ||
    document.webkitFullscreenEnabled ||
    document.mozFullScreenEnabled ||
    document.msFullscreenEnabled
  )
}

export const isFullscreen = () => {
  return (
    document.fullscreenElement ||
    document.webkitFullscreenElement ||
    document.mozFullScreenElement ||
    document.msFullscreenElement
  )
}

export const addFSEventListener = handler => {
  document.addEventListener("fullscreenchange", handler)
  document.addEventListener("webkitfullscreenchange", handler)
  document.addEventListener("mozfullscreenchange", handler)
  document.addEventListener("MSFullscreenChange", handler)
}

export const removeFSEventListener = handler => {
  document.removeEventListener("fullscreenchange", handler)
  document.removeEventListener("webkitfullscreenchange", handler)
  document.removeEventListener("mozfullscreenchange", handler)
  document.removeEventListener("MSFullscreenChange", handler)
}
