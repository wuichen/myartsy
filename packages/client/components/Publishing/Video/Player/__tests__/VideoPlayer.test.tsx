import { mount } from "enzyme"
import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import track from "react-tracking"
import * as FullscreenHelpers from "../FullscreenHelpers"
import { VideoPlayer } from "../VideoPlayer"

jest.useFakeTimers()

describe("VideoPlayer", () => {
  const trackEvent = jest.fn()

  const getWrapper = () => {
    return mount<VideoPlayer>(
      <VideoPlayer
        url="http://files.artsy.net/videos/placeholder.mp4"
        tracking={{
          trackEvent,
          getTrackingData: jest.fn(),
        }}
      />
    )
  }

  beforeEach(() => {
    ;(FullscreenHelpers as any).addFSEventListener = jest.fn()
    ;(FullscreenHelpers as any).requestFullscreen = jest.fn()
    ;(FullscreenHelpers as any).exitFullscreen = jest.fn()
    ;(document as any).fullscreenEnabled = true
  })

  it("matches the snapshot", () => {
    const videoPlayer = renderer
      .create(
        <VideoPlayer
          url="http://files.artsy.net/videos/placeholder.mp4"
          title="The Video Title"
        />
      )
      .toJSON()
    expect(videoPlayer).toMatchSnapshot()
  })

  describe("Player controls", () => {
    it("#setDuration", () => {
      const videoPlayer = getWrapper()
      videoPlayer.instance().setDuration({
        target: {
          duration: 10000,
        },
      })
      expect(videoPlayer.state("duration")).toBe(10000)
    })

    it("#updateTime", () => {
      const videoPlayer = getWrapper()
      videoPlayer.instance().updateTime({
        target: {
          currentTime: 20000,
        },
      })
      expect(videoPlayer.state("currentTime")).toBe(20000)
    })

    it("#togglePlay - pauses video if its playing", () => {
      const videoPlayer = getWrapper()
      videoPlayer.setState({ isPlaying: true })
      videoPlayer.instance().togglePlay()
      expect(videoPlayer.state("isPlaying")).toBe(false)
    })

    it("#togglePlay - plays video if its paused", () => {
      const videoPlayer = getWrapper()
      videoPlayer.setState({ isPlaying: false })
      videoPlayer.instance().togglePlay()
      expect(videoPlayer.state("isPlaying")).toBe(true)
    })

    it("#toggleMute - mutes video if its unmuted", () => {
      const videoPlayer = getWrapper()
      videoPlayer.setState({ isMuted: false })
      videoPlayer.instance().toggleMute()
      expect(videoPlayer.state("isMuted")).toBe(true)
    })

    it("#toggleMute - mutes video if its unmuted", () => {
      const videoPlayer = getWrapper()
      videoPlayer.setState({ isMuted: true })
      videoPlayer.instance().toggleMute()
      expect(videoPlayer.state("isMuted")).toBe(false)
    })

    it("#toggleFullscreen - enters fullscreen", () => {
      const videoPlayer = getWrapper()
      videoPlayer.instance().toggleFullscreen()
      expect(FullscreenHelpers.requestFullscreen).toBeCalled()
    })

    it("#toggleFullscreen - exits fullscreen", () => {
      const videoPlayer = getWrapper()
      ;(FullscreenHelpers as any).isFullscreen = () => true
      videoPlayer.instance().toggleFullscreen()
      expect(FullscreenHelpers.exitFullscreen).toBeCalled()
    })

    it("#seekTo", () => {
      const videoPlayer = getWrapper()
      videoPlayer.instance().seekTo(5000)
      expect(videoPlayer.state("currentTime")).toBe(5000)
      expect(videoPlayer.instance().video.currentTime).toBe(5000)
    })

    it("#pause", () => {
      const videoPlayer = getWrapper()
      jest.spyOn(videoPlayer.instance().video, "pause")
      videoPlayer.instance().pause()
      expect(videoPlayer.state("isPlaying")).toBe(false)
      expect(videoPlayer.instance().video.pause).toBeCalled()
    })

    it("#play", () => {
      const videoPlayer = getWrapper()
      jest.spyOn(videoPlayer.instance().video, "play")
      videoPlayer.instance().play()
      expect(videoPlayer.state("isPlaying")).toBe(true)
      expect(videoPlayer.instance().video.play).toBeCalled()
    })

    it("fades out controls after 3 seconds of idle time", () => {
      const videoPlayer = getWrapper()
      jest.runTimersToTime(4000)
      expect(videoPlayer.state("showControls")).toBe(false)
    })

    it("controls stay visible under 3 seconds of idle time", () => {
      const videoPlayer = getWrapper()
      jest.runTimersToTime(2000)
      expect(videoPlayer.state("showControls")).toBe(true)
    })

    it("fades in controls on mousemove", () => {
      const videoPlayer = getWrapper()
      videoPlayer.simulate("mousemove")
      expect(videoPlayer.state("showControls")).toBe(true)
      expect(videoPlayer.state("idleTime")).toBe(0)
    })

    it("fades in controls on keypress", () => {
      const videoPlayer = getWrapper()
      videoPlayer.simulate("keypress")
      expect(videoPlayer.state("showControls")).toBe(true)
      expect(videoPlayer.state("idleTime")).toBe(0)
    })

    it("fades in controls on touchstart", () => {
      const videoPlayer = getWrapper()
      videoPlayer.simulate("touchstart")
      expect(videoPlayer.state("showControls")).toBe(true)
      expect(videoPlayer.state("idleTime")).toBe(0)
    })
  })

  describe("Analytics", () => {
    it("tracks fullscreen button", () => {
      const videoPlayer = getWrapper()
      videoPlayer.instance().toggleFullscreen()
      expect(track).toBeCalled()
    })

    it("tracks duration", () => {
      const videoPlayer = getWrapper()
      Object.defineProperty(videoPlayer.instance().video, "duration", {
        writable: true,
      })
      videoPlayer.instance().video.currentTime = 3
      ;(videoPlayer.instance().video as any).duration = 10
      videoPlayer.instance().trackProgress()
      expect(trackEvent).toBeCalledWith({
        action: "Video duration",
        label: "Video duration",
        percent_complete: 25,
      })
    })

    it("tracks seconds passed", () => {
      const videoPlayer = getWrapper()
      videoPlayer.instance().video.currentTime = 3
      videoPlayer.instance().trackProgress()
      expect(trackEvent).toBeCalledWith({
        action: "Video seconds",
        label: "Video seconds",
        seconds_complete: 3,
      })
    })
  })
})
