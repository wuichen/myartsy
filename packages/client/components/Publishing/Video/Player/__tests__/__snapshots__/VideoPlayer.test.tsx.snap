// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer matches the snapshot 1`] = `
.c14 {
  width: 28px;
  height: 28px;
}

.c12 {
  height: 28px;
  cursor: pointer;
}

.c7 {
  width: 32px;
  height: 32px;
}

.c6 {
  height: 28px;
  width: 28px;
  cursor: pointer;
}

.c15 {
  width: 100%;
  background-size: NaN% 100%;
  height: 18px;
  margin-top: 15px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
}

.c15::-webkit-slider-thumb {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c15::-ms-track {
  width: 100%;
  height: 2px;
  cursor: pointer;
  background: transparent;
  border: 0px;
  color: transparent;
}

.c15::-webkit-slider-runnable-track {
  height: 2px;
  cursor: pointer;
  -webkit-animate: 0.2s;
  animate: 0.2s;
  background-color: white;
  border: 0px;
}

.c15::-moz-range-track {
  height: 2px;
  cursor: pointer;
  -webkit-animate: 0.2s;
  animate: 0.2s;
  background-color: white;
  border: 0px;
}

.c15::-webkit-slider-thumb {
  color: white;
  border: 0px;
  height: 12px;
  width: 12px;
  border-radius: 50px;
  background: white;
  cursor: pointer;
  margin-top: -5px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c15::-moz-range-thumb {
  color: white;
  box-shadow: 0px 0px 0px white;
  border: 0px;
  height: 12px;
  width: 12px;
  border-radius: 50px;
  background: white;
  cursor: pointer;
  margin-top: -5px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c15::-ms-fill-lower {
  background: white;
  border: 0px;
}

.c15::-ms-fill-upper {
  background: white;
  border: 0px;
}

.c15::-ms-thumb {
  color: white;
  box-shadow: 0px 0px 0px white;
  border: 0px;
  height: 12px;
  width: 12px;
  border-radius: 50px;
  background: white;
  cursor: pointer;
  margin-top: -5px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c15:focus::-webkit-slider-runnable-track {
  background: white;
}

.c15:focus {
  outline: none;
}

.c10 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 23px;
  line-height: 1.5em;
  -webkit-font-smoothing: antialiased;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c3 {
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  border: 1px solid white;
  border-radius: 2px;
  padding: 20px;
  margin: 20px;
  opacity: 0;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c3 .c9 {
  margin-right: 30px;
}

.c3 .c11 {
  margin-right: 20px;
}

.c3 .c13 {
  cursor: pointer;
}

.c8 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 23px;
  line-height: 1.5em;
  -webkit-font-smoothing: antialiased;
  margin-left: 20px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  margin: 0 auto;
  max-width: 1200px;
  width: 100%;
}

.c0 {
  width: 100%;
  height: 100%;
  position: relative;
  background-color: black;
}

.c0 video {
  object-fit: contain;
  width: 100%;
  height: 100%;
}

.c0 .c2 {
  opacity: 1;
}

<div
  className="c0"
>
  <video
    muted={false}
    playsInline={true}
    src="http://files.artsy.net/videos/placeholder.mp4"
  />
  <div
    className="c1"
  >
    <div
      className="c2 c3"
    >
      <div
        className="c4"
      >
        <div
          className="c5"
        >
          <div
            className="c6"
            onClick={[Function]}
          >
            <svg
              className="c7"
              version="1.1"
              viewBox="0 0 40 56"
              x="0px"
              xmlns="http://www.w3.org/2000/svg"
              y="0px"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeWidth="1"
              >
                <polygon
                  fill="white"
                  points="0 0 0 56 39.3600006 28"
                />
              </g>
            </svg>
          </div>
          <div
            className="c8"
          >
            The Video Title
          </div>
        </div>
        <div
          className="c5"
        >
          <div
            className="c9 c10"
          >
            00:00
             / 
            00:00
          </div>
          <div
            className="c11 c12"
            onClick={[Function]}
          >
            <svg
              height="30px"
              width="30px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="white"
              >
                <path
                  d="M19.1285714,15.3428571 L19.1285714,28.7571429 L11.8285714,21.4714286 L0.757142857,21.4714286 L0.757142857,8.04285714 L11.8285714,8.04285714 L19.1285714,0.757142857 L19.1285714,15.3428571 Z"
                />
              </g>
            </svg>
          </div>
          <svg
            className="c13 c14"
            height="32px"
            onClick={[Function]}
            viewBox="0 0 28 28"
            width="32px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fillRule="evenodd"
              transform="translate(-1151 -22)"
            >
              <rect
                height="92"
                rx="4"
                stroke="white"
                width="1200"
              />
              <path
                d="M1150 21h30v30h-30z"
              />
              <g
                fill="#FFF"
              >
                <path
                  d="M1151 40h2v10h-2z"
                />
                <path
                  d="M1161 48v2h-10v-2zM1161 22v2h-10v-2z"
                />
                <path
                  d="M1153 32h-2V22h2z"
                />
                <g>
                  <path
                    d="M1179 32h-2V22h2z"
                  />
                  <path
                    d="M1169 24v-2h10v2z"
                  />
                </g>
                <g>
                  <path
                    d="M1169 50v-2h10v2z"
                  />
                  <path
                    d="M1177 40h2v10h-2z"
                  />
                </g>
              </g>
              <path
                d="M1157 28h16v16h-16z"
                fill="#FFF"
              />
            </g>
          </svg>
        </div>
      </div>
      <input
        className="c15"
        max={0}
        onChange={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        step="any"
        type="range"
        value={0}
      />
    </div>
  </div>
</div>
`;
